packages/vm/src/imports.rs

```diff
@@ -37,172 +33,222 @@ const MAX_LENGTH_QUERY_CHAIN_REQUEST: usize = 64 * KI;
     result?;
 
     Ok(())
 }
 
 pub fn do_canonicalize_address<A: Api, S: Storage, Q: Querier>(
     api: A,
-    ctx: &mut Ctx,
+    env: &Env<S, Q>,
     source_ptr: u32,
     destination_ptr: u32,
 ) -> VmResult<u32> {
-    let source_data = read_region(ctx, source_ptr, MAX_LENGTH_HUMAN_ADDRESS)?;
+    let source_data = read_region(&env.memory(), source_ptr, MAX_LENGTH_HUMAN_ADDRESS)?;
     if source_data.is_empty() {
-        return Ok(write_to_contract::<S, Q>(ctx, b"Input is empty")?);
+        return Ok(write_to_contract::<S, Q>(env, b"Input is empty")?);
     }
 
     let source_string = match String::from_utf8(source_data) {
         Ok(s) => s,
-        Err(_) => return Ok(write_to_contract::<S, Q>(ctx, b"Input is not valid UTF-8")?),
+        Err(_) => return Ok(write_to_contract::<S, Q>(env, b"Input is not valid UTF-8")?),
     };
     let human: HumanAddr = source_string.into();
 
     let (result, gas_info) = api.canonical_address(&human);
-    process_gas_info::<S, Q>(ctx, gas_info)?;
+    process_gas_info::<S, Q>(env, gas_info)?;
     match result {
         Ok(canonical) => {
-            write_region(ctx, destination_ptr, canonical.as_slice())?;
+            write_region(&env.memory(), destination_ptr, canonical.as_slice())?;
             Ok(0)
         }
         Err(BackendError::UserErr { msg, .. }) => {
-            Ok(write_to_contract::<S, Q>(ctx, msg.as_bytes())?)
+            Ok(write_to_contract::<S, Q>(env, msg.as_bytes())?)
         }
         Err(err) => Err(VmError::from(err)),
     }
 }
 
 pub fn do_humanize_address<A: Api, S: Storage, Q: Querier>(
     api: A,
-    ctx: &mut Ctx,
+    env: &Env<S, Q>,
     source_ptr: u32,
     destination_ptr: u32,
 ) -> VmResult<u32> {
-    let canonical = Binary(read_region(ctx, source_ptr, MAX_LENGTH_CANONICAL_ADDRESS)?);
+    let canonical = Binary(read_region(
+        &env.memory(),
+        source_ptr,
+        MAX_LENGTH_CANONICAL_ADDRESS,
+    )?);
 
     let (result, gas_info) = api.human_address(&CanonicalAddr(canonical));
-    process_gas_info::<S, Q>(ctx, gas_info)?;
+    process_gas_info::<S, Q>(env, gas_info)?;
     match result {
         Ok(human) => {
-            write_region(ctx, destination_ptr, human.as_str().as_bytes())?;
+            write_region(&env.memory(), destination_ptr, human.as_str().as_bytes())?;
             Ok(0)
         }
         Err(BackendError::UserErr { msg, .. }) => {
-            Ok(write_to_contract::<S, Q>(ctx, msg.as_bytes())?)
+            Ok(write_to_contract::<S, Q>(env, msg.as_bytes())?)
         }
         Err(err) => Err(VmError::from(err)),
     }
 }
 
 /// Prints a debug message to console.
 /// This does not charge gas, so debug printing should be disabled when used in a blockchain module.
-pub fn print_debug_message(ctx: &mut Ctx, message_ptr: u32) -> VmResult<()> {
-    let message_data = read_region(ctx, message_ptr, MAX_LENGTH_DEBUG)?;
+pub fn print_debug_message<S: Storage, Q: Querier>(
```

Comment:
Wouldn't this alos be a candidate for a "native_*" function?

Reply:
It captures `print_debug` in 

```rust
        env_imports.insert(
            "debug",
            Function::new_with_env(store, &i32_to_void, env.clone(), move |env, args| {
                let message_ptr = ref_to_u32(&args[0])?;
                if print_debug {
                    print_debug_message(&env, message_ptr)?;
                }
                Ok(vec![])
            }),
        );
```

But this can go into Env.

Reply:
Done in e657e59be66b2c6764eae2c24b1cabeabff45010